File: games/connect4/board.py
"""
A Connect 4 position, and the rules for getting from one to the next.

The position is two integers, one per player, over the sentinel-padded layout described in
constants.py. Two boards rather than the position/mask pairing a lot of Connect 4 engines use,
for two reasons: making and unmaking a move becomes `self.discs[player] ^= bit`, which is its
own inverse and so cannot be got wrong by undoing in the wrong order; and the evaluation wants
to ask about one player's discs on their own anyway.
"""

from __future__ import annotations

from typing import Dict, Iterable, Iterator, List, Optional

from games.base import DRAW, GameState, Outcome, win
from games.connect4.bitboard import (
    Bitboard,
    bit_count,
    drops,
    index,
    is_win,
    landing_square,
)
from games.connect4.constants import (
    CENTRE_FIRST,
    COLS,
    COLUMN_MASKS,
    DISC_ICONS,
    FULL_BOARD,
    RED,
    ROWS,
    STRIDE,
    YELLOW,
    Disc,
)
from games.connect4.evaluation import weighted_eval


class IllegalMove(Exception):
    """A column that is not on the board, or one that is already full."""


class Connect4(GameState):
    # Seven moves at the root, and an evaluation cheap enough that a node costs less than
    # handing it to another process would. Starting a pool here would lose every time.
    PARALLEL_ROOT = False

    DEFAULT_EVAL = staticmethod(weighted_eval)

    def __init__(self, columns: Iterable[int] = ()) -> None:
        """A position, reached by playing `columns` from an empty board."""
        self.discs: Dict[Disc, Bitboard] = {YELLOW: 0, RED: 0}
        self.turn: Disc = YELLOW
        self.move_stack: List[Bitboard] = []

        for column in columns:
            if column not in self.legal_moves:
                raise IllegalMove(f'column {column} cannot be played here')
            self.make_move(column)

    # ---- position ----------------------------------------------------------------------

    @property
    def occupied(self) -> Bitboard:
        """
        Every cell holding a disc.

        Derived rather than cached. Caching it means make_move and unmake_move have to keep it
        in step, and a cache that is right almost always is worse than no cache at all here -
        move generation and win detection would both quietly lie. Measure before adding one.
        """
        return self.discs[YELLOW] | self.discs[RED]

    @property
    def legal_moves(self) -> Iterator[int]:
        """
        Every column with room in it, nearest the middle first.

        The order is deliberate and load-bearing. `ai.search` iterates moves in generation
        order and breaks ties by it, so centre-first generation is what makes alpha-beta's
        pruning pay - best-move-first takes the cost from about 7^d to 7^(d/2), and worst-first
        gains nothing at all. It also means a search with nothing to choose between takes the
        centre column, which happens to be the right opening move.
        """
        landing = drops(self.occupied)
        return (column for column in CENTRE_FIRST if landing & COLUMN_MASKS[column])

    @property
    def signature(self) -> str:
        """
        The two disc masks, which are the whole of the position.

        Connect 4 carries none of the state chess hides behind an identical-looking board - no
        castling rights, no en passant square, no clock - so nothing is lost by not printing it.
        Overridden only because the inherited default renders the whole grid, and the
        conformance suite asks for this once per move of every position it walks.
        """
        return f'{self.discs[YELLOW]}/{self.discs[RED]}'

    def copy(self) -> 'Connect4':
        """
        An independent board at the same position.

        The move stack comes too. Chess drops its history here because rebuilding from a FEN is
        cheaper than copying it, but ours is a list of small ints and copying it costs nothing,
        so a copy stays undoable to the same depth the original was.
        """
        clone = Connect4()
        clone.discs[YELLOW] = self.discs[YELLOW]
        clone.discs[RED] = self.discs[RED]
        clone.turn = self.turn
        clone.move_stack = list(self.move_stack)
        return clone

    # ---- playing -----------------------------------------------------------------------

    def make_move(self, column: int) -> None:
        """
        Drops a disc into `column` and hands over the turn.

        Assumes the column is legal - this is the search's hot path, and `legal_moves` is where
        legality is decided. A full column yields a landing square of 0, so an illegal move here
        is a silent no-op on the boards and a corrupt turn, which is exactly what the
        `landing_square` assertion is for when running under -O0.

        `^=` rather than `|=`, throughout and on purpose. Addition and OR both happen to work
        while the target cell is provably empty and fail silently when it is not; XOR is its own
        inverse, so unmake_move is the same line again.
        """
        bit = landing_square(self.occupied, column)
        assert bit, f'column {column} is full'

        self.discs[self.turn] ^= bit
        self.move_stack.append(bit)
        self.turn = not self.turn

    def unmake_move(self) -> None:
        """Takes the last disc back off the board."""
        bit = self.move_stack.pop()
        self.turn = not self.turn
        self.discs[self.turn] ^= bit

    # ---- results -----------------------------------------------------------------------

    @property
    def outcome(self) -> Optional[Outcome]:
        """
        The winner, if the last move made one.

        This is the property chess does not have. Chess ends by a player running out of moves,
        so it can leave this as the inherited None; Connect 4 is won with the board half empty
        and legal moves still on offer, so without this the search would play straight through a
        finished game.

        Only one board is tested, because only one player can have just won: the one who moved.
        That halves the cost of the property the search touches at every single node. It is
        sound because no constructor here can build a position where the player *to* move
        already has four - `from_diagram` rejects those, and play cannot reach one.

        At the start the mover is Red, whose board is 0, and an empty board has no run, so no
        special case is needed for a game that has not begun.
        """
        mover = not self.turn
        return win(mover) if is_win(self.discs[mover]) else None

    @property
    def outcome_without_moves(self) -> Outcome:
        """Nowhere to play means the grid is full, and a full grid nobody won is a draw."""
        return DRAW

    # ---- reading and writing positions -------------------------------------------------

    @classmethod
    def from_diagram(cls, diagram: str) -> 'Connect4':
        """
        A position from a picture of one, top row first, for tests that want to be read.

        `.` is empty, `Y` and `R` are discs:

            Connect4.from_diagram('''
                .......
                .......
                .......
                .......
                ...R...
                ..RYY..
            ''')

        This is the one sanctioned way to set the boards other than by playing, so it carries
        the checks that playing gets for free. A diagram with a floating disc, with the wrong
        number of each colour, or in which the player *to move* has already won would break the
        assumption `outcome` rests on, and would do it invisibly: `result` delegates to
        `outcome`, so the conformance suite could not catch it either.
        """
        rows = [line.strip() for line in diagram.strip().splitlines()]
        rows = [row for row in rows if row]
        if len(rows) != ROWS:
            raise ValueError(f'expected {ROWS} rows, got {len(rows)}')

        board = cls()
        for height, line in enumerate(rows):
            if len(line) != COLS:
                raise ValueError(f'expected {COLS} columns, got {len(line)} in {line!r}')

            row = ROWS - 1 - height  # Diagrams read downwards; rows count upwards
            for column, char in enumerate(line):
                if char in 'Yy':
                    board.discs[YELLOW] |= 1 << index(column, row)
                elif char in 'Rr':
                    board.discs[RED] |= 1 << index(column, row)
                elif char not in '.-_ ':
                    raise ValueError(f'unrecognised cell {char!r}')

        board._validate()
        return board

    def _validate(self) -> None:
        """Checks a hand-built position is one that play could actually have reached."""
        yellow, red = self.discs[YELLOW], self.discs[RED]

        if yellow & red:
            raise ValueError('a cell holds two discs')
        if (yellow | red) & ~FULL_BOARD:
            raise ValueError('a disc is off the board')

        played = bit_count(yellow), bit_count(red)
        if played[0] not in (played[1], played[1] + 1):
            raise ValueError(f'Yellow has played {played[0]} discs and Red {played[1]}')
        self.turn = YELLOW if played[0] == played[1] else RED

        occupied = self.occupied
        for column in range(COLS):
            # Discs stack from the bottom, so a column read down to bit 0 must be a solid run.
            stack = (occupied & COLUMN_MASKS[column]) >> (column * STRIDE)
            if stack != (1 << bit_count(stack)) - 1:
                raise ValueError(f'column {column} has a disc floating above a gap')

        if is_win(self.discs[self.turn]):
            raise ValueError('the player to move has already won')

    @property
    def columns_played(self) -> List[int]:
        """The columns of the moves on the stack, in the order they were played."""
        return [bit.bit_length() // STRIDE for bit in self.move_stack]

    def parse_move(self, text: str) -> int:
        """A column number, as typed. Columns are numbered from 0, as the board prints them."""
        try:
            column = int(text.strip())
        except ValueError:
            raise ValueError(f'{text!r} is not a column number') from None

        if column not in self.legal_moves:
            if 0 <= column < COLS:
                raise ValueError(f'column {column} is full')
            raise ValueError(f'there is no column {column}')
        return column

    def __str__(self) -> str:
        """The board as a picture, top row first, with the column numbers under it."""
        lines = []
        for row in range(ROWS - 1, -1, -1):
            cells = ''
            for column in range(COLS):
                bit = 1 << index(column, row)
                if self.discs[YELLOW] & bit:
                    cells += f'[{DISC_ICONS[YELLOW]}]'
                elif self.discs[RED] & bit:
                    cells += f'[{DISC_ICONS[RED]}]'
                else:
                    cells += '[ ]'
            lines.append(cells)
        lines.append(' ' + '  '.join(str(column) for column in range(COLS)) + ' ')
        return '\n' + '\n'.join(lines)

File: rust/crates/c4-core/tests/board.rs
//! The board against the counts and invariants `tests/connect4/` pins for the Python engine.

use c4_core::bitboard::{drops, index, is_win, landing_square, runs};
use c4_core::constants::{
    BOTTOM_ROW, CENTRE_FIRST, COLS, COLUMN_MASKS, CONNECT, DIRECTIONS, FULL_BOARD, ROWS, STRIDE,
};
use c4_core::encode::{legal_mask, planes, PLANE_BYTES, PLANE_SIZE};
use c4_core::Connect4;

/// A deterministic stand-in for a random number generator, so a fuzz test is reproducible.
struct Lcg(u64);

impl Lcg {
    fn next(&mut self, bound: usize) -> usize {
        self.0 = self.0.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        ((self.0 >> 33) as usize) % bound
    }
}

// ---- the masks, re-derived independently ------------------------------------------------

#[test]
fn the_board_is_every_playable_cell_and_nothing_else() {
    let mut expected = 0u64;
    for column in 0..COLS {
        for row in 0..ROWS {
            expected |= 1 << index(column, row);
        }
    }
    assert_eq!(FULL_BOARD, expected);
    assert_eq!(FULL_BOARD.count_ones(), ROWS * COLS);
}

#[test]
fn the_sentinel_row_is_never_part_of_the_board() {
    for column in 0..COLS {
        assert_eq!(FULL_BOARD & (1 << index(column, ROWS)), 0);
    }
}

#[test]
fn column_masks_tile_the_board() {
    let mut union = 0u64;
    for mask in COLUMN_MASKS {
        assert_eq!(union & mask, 0, "columns overlap");
        union |= mask;
    }
    assert_eq!(union, FULL_BOARD);
}

#[test]
fn the_bottom_row_is_one_cell_per_column() {
    assert_eq!(BOTTOM_ROW.count_ones(), COLS);
    for column in 0..COLS {
        assert_eq!(BOTTOM_ROW & COLUMN_MASKS[column as usize], 1 << index(column, 0));
    }
}

#[test]
fn moves_are_offered_centre_first() {
    assert_eq!(CENTRE_FIRST, [3, 2, 4, 1, 5, 0, 6]);
}

// ---- win detection ----------------------------------------------------------------------

/// Four in a row found by scanning the grid, sharing nothing with the shift chains.
fn naive_is_win(position: u64) -> bool {
    let steps: [(i32, i32); 4] = [(0, 1), (1, 0), (1, 1), (1, -1)];
    for column in 0..COLS as i32 {
        for row in 0..ROWS as i32 {
            for (dc, dr) in steps {
                let mut found = 0;
                for step in 0..CONNECT as i32 {
                    let (c, r) = (column + dc * step, row + dr * step);
                    if c < 0 || c >= COLS as i32 || r < 0 || r >= ROWS as i32 {
                        break;
                    }
                    if position >> index(c as u32, r as u32) & 1 == 0 {
                        break;
                    }
                    found += 1;
                }
                if found == CONNECT as i32 {
                    return true;
                }
            }
        }
    }
    false
}

#[test]
fn every_line_of_four_is_a_win() {
    let steps: [(i32, i32); 4] = [(0, 1), (1, 0), (1, 1), (1, -1)];
    let mut lines = 0;
    for column in 0..COLS as i32 {
        for row in 0..ROWS as i32 {
            for (dc, dr) in steps {
                let cells: Vec<(i32, i32)> = (0..CONNECT as i32)
                    .map(|step| (column + dc * step, row + dr * step))
                    .collect();
                if cells.iter().any(|&(c, r)| {
                    c < 0 || c >= COLS as i32 || r < 0 || r >= ROWS as i32
                }) {
                    continue;
                }
                let position = cells
                    .iter()
                    .fold(0u64, |bb, &(c, r)| bb | 1 << index(c as u32, r as u32));
                assert!(is_win(position), "missed a line at {column},{row} {dc},{dr}");
                lines += 1;
            }
        }
    }
    assert_eq!(lines, 69, "the 7x6 board has 69 lines of four");
}

#[test]
fn a_run_cannot_wrap_between_columns() {
    // Four cells that are consecutive in the *bit* layout but span two columns in every direction.
    for delta in DIRECTIONS {
        for start in 0..(COLS * STRIDE - CONNECT * delta) {
            let position = (0..CONNECT).fold(0u64, |bb, step| bb | 1 << (start + step * delta));
            let inside = position & FULL_BOARD == position;
            assert_eq!(
                is_win(position & FULL_BOARD),
                inside && naive_is_win(position & FULL_BOARD),
                "delta {delta} from bit {start}"
            );
        }
    }
}

#[test]
fn win_detection_agrees_with_a_grid_scan_over_random_boards() {
    let mut rng = Lcg(20260814);
    for _ in 0..20_000 {
        let mut board = Connect4::new();
        let mut moves = [0u8; COLS as usize];
        loop {
            let count = board.legal_moves(&mut moves);
            if count == 0 || board.outcome().is_some() {
                break;
            }
            board.make_move(moves[rng.next(count)]);
        }
        for player in [true, false] {
            assert_eq!(is_win(board.discs(player)), naive_is_win(board.discs(player)));
        }
    }
}

#[test]
fn runs_travels_exactly_length_minus_one() {
    // A vertical line of three has its lowest cell marked at length 3 and nothing at length 4.
    let three = 1 | 1 << 1 | 1 << 2;
    assert_eq!(runs(three, 1, 3), 1);
    assert_eq!(runs(three, 1, 4), 0);
    assert_eq!(runs(three | 1 << 3, 1, 4), 1);
    assert_eq!(runs(three, 1, 2), 1 | 1 << 1);
}

// ---- move generation and make/unmake -----------------------------------------------------

#[test]
fn a_carry_finds_the_landing_square_in_every_column() {
    let mut board = Connect4::new();
    for column in 0..COLS {
        for row in 0..ROWS {
            assert_eq!(landing_square(board.occupied(), column), 1 << index(column, row));
            board.make_move(column as u8);

            // A move somewhere else, taken straight back: unmaking must leave this column's
            // landing square where it was. There is nowhere spare once the board is nearly full,
            // and the assertion above is the point of the loop either way.
            let spare = (0..COLS)
                .find(|&other| other != column && landing_square(board.occupied(), other) != 0);
            if let Some(other) = spare {
                board.make_move(other as u8);
                board.unmake_move();
            }
        }
        assert_eq!(landing_square(board.occupied(), column), 0, "full column still offered");
    }
    assert_eq!(drops(board.occupied()), 0);
    assert!(!board.has_moves());
}

#[test]
fn column_zero_alone_is_still_a_move() {
    // The trap `games/base.py` documents: move 0 is legal and falsy in Python.
    let mut board = Connect4::new();
    for column in 1..COLS {
        for _ in 0..ROWS {
            board.make_move(column as u8);
        }
    }
    let mut moves = [0u8; COLS as usize];
    assert_eq!(board.legal_moves(&mut moves), 1);
    assert_eq!(moves[0], 0);
    assert!(board.has_moves());
    assert!(!board.is_game_over() || board.outcome().is_some());
}

#[test]
fn unmake_restores_the_position_exactly() {
    let mut rng = Lcg(1);
    let mut board = Connect4::new();
    let mut moves = [0u8; COLS as usize];

    for _ in 0..50_000 {
        let count = board.legal_moves(&mut moves);
        if count == 0 || board.outcome().is_some() {
            board = Connect4::new();
            continue;
        }
        let before = (board.discs(true), board.discs(false), board.turn(), board.ply());
        let column = moves[rng.next(count)];

        board.make_move(column);
        board.unmake_move();
        assert_eq!(
            (board.discs(true), board.discs(false), board.turn(), board.ply()),
            before
        );
        board.make_move(column);
    }
}

// ---- perft -------------------------------------------------------------------------------

/// Leaves below this position, counting a finished game as having none. The horizon is checked
/// before the outcome, so a win at exactly `depth` counts as one leaf - see `ai/perft.py`.
fn perft(board: &mut Connect4, depth: u32) -> u64 {
    if depth == 0 {
        return 1;
    }
    if board.outcome().is_some() {
        return 0;
    }

    let mut moves = [0u8; COLS as usize];
    let count = board.legal_moves(&mut moves);
    let mut total = 0;
    for &column in moves.iter().take(count) {
        board.make_move(column);
        total += perft(board, depth - 1);
        board.unmake_move();
    }
    total
}

#[test]
fn perft_matches_the_pinned_counts() {
    let expected = [7, 49, 343, 2_401, 16_807, 117_649, 823_536];
    let mut board = Connect4::new();
    for (depth, &nodes) in expected.iter().enumerate() {
        assert_eq!(perft(&mut board, depth as u32 + 1), nodes, "depth {}", depth + 1);
    }
}

#[test]
#[ignore = "about a second; the shorter depths catch the same faults"]
fn perft_eight_matches_the_pinned_count() {
    assert_eq!(perft(&mut Connect4::new(), 8), 5_673_234);
}

// ---- the encoder --------------------------------------------------------------------------

#[test]
fn plane_zero_is_always_the_mover() {
    let mut board = Connect4::from_columns(&[3, 3, 4]);
    let mut mine = [0u8; PLANE_BYTES];
    planes(&board, &mut mine);

    // Yellow has two discs on the board and is to move again after four plies.
    assert_eq!(mine[..PLANE_SIZE].iter().map(|&c| c as u32).sum::<u32>(), 1);
    assert_eq!(mine[PLANE_SIZE..].iter().map(|&c| c as u32).sum::<u32>(), 2);

    board.make_move(0);
    let mut theirs = [0u8; PLANE_BYTES];
    planes(&board, &mut theirs);
    assert_eq!(theirs[..PLANE_SIZE], mine[PLANE_SIZE..]);
}

#[test]
fn row_zero_is_the_bottom_of_the_board() {
    let board = Connect4::from_columns(&[6]);
    let mut cells = [0u8; PLANE_BYTES];
    planes(&board, &mut cells);

    // The mover is now Red, so Yellow's lone disc is in plane 1, bottom row, column 6.
    assert_eq!(cells[PLANE_SIZE + (COLS - 1) as usize], 1);
    assert_eq!(cells.iter().map(|&c| c as u32).sum::<u32>(), 1);
}

#[test]
fn the_legal_mask_is_the_legal_moves() {
    let mut rng = Lcg(7);
    let mut board = Connect4::new();
    let mut moves = [0u8; COLS as usize];

    for _ in 0..5_000 {
        let count = board.legal_moves(&mut moves);
        if count == 0 || board.outcome().is_some() {
            board = Connect4::new();
            continue;
        }
        let mut mask = [false; COLS as usize];
        legal_mask(&board, &mut mask);
        assert_eq!(mask.iter().filter(|&&legal| legal).count(), count);
        for &column in moves.iter().take(count) {
            assert!(mask[column as usize]);
        }
        board.make_move(moves[rng.next(count)]);
    }
}
